tests/unit/imports/server/publishJoinedQuery.ts

```diff
@@ -0,0 +1,174 @@
+
+  stop() { /* noop */ }
+
+  error(_error: Error) { /* noop */ }
+
+  unblock() { /* noop */ }
+
+  userId: string | null = null;
+
+  get connection(): Meteor.Connection {
+    throw new Error('Method not implemented.');
+  }
+}
+
+const safeDrop = async (collection: Mongo.Collection<any>) => {
+  try {
+    await collection.dropCollectionAsync();
+  } catch (e) {
+    if (!(e instanceof MongoError) || e.code !== 26) {
+      throw e;
+    }
+  }
+};
+
+describe('publishJoinedQuery', function () {
+  beforeEach(async function () {
+    await Promise.all([
+      safeDrop(Hunts),
+      safeDrop(Puzzles),
+      safeDrop(Guesses),
+      safeDrop(Tags),
+    ]);
+    await makeFixtureHunt(Random.id());
+  });
+
+  it('can follow a string foreign key', function () {
+    const sub = new StubSubscription();
+    after(() => sub.stopHooks.forEach((hook) => hook()));
+
+    const guessId = 'obeeKs3ZEkBe3ykeg';
+    const puzzleId = 'fXchzrh8X9EoSZu6k';
+
+    publishJoinedQuery(sub, {
+      model: Guesses,
+      foreignKeys: [{
+        field: 'puzzle',
+        join: {
+          model: Puzzles,
+        },
+      }],
+    }, { _id: guessId });
+
+    assert.sameMembers([...sub.data.keys()], [Puzzles._name, Guesses._name]);
+
+    const guessCollection = sub.data.get(Guesses._name)!;
+    assert.sameMembers([...guessCollection.keys()], [guessId]);
+
+    const puzzleCollection = sub.data.get(Puzzles._name)!;
+    assert.sameMembers([...puzzleCollection.keys()], [puzzleId]);
+  });
+
+  it('can follow an array foreign key', function () {
+    const sub = new StubSubscription();
+    after(() => sub.stopHooks.forEach((hook) => hook()));
+
+    const puzzleId = 'fXchzrh8X9EoSZu6k';
+    const tagIds = ['o5JdfTizW4tGwhRnP', 'QeJLufdCqv7rMSSbS'];
+
+    publishJoinedQuery(sub, {
+      model: Puzzles,
+      foreignKeys: [{
+        field: 'tags',
+        join: {
+          model: Tags,
+        },
+      }],
+    }, { _id: puzzleId });
+
+    assert.sameMembers([...sub.data.keys()], [Puzzles._name, Tags._name]);
+
+    const puzzleCollection = sub.data.get(Puzzles._name)!;
+    assert.sameMembers([...puzzleCollection.keys()], [puzzleId]);
+
+    const tagCollection = sub.data.get(Tags._name)!;
+    assert.sameMembers([...tagCollection.keys()], tagIds);
+  });
+
+  it('updates if foreign keys change', async function () {
+    const sub = new StubSubscription();
+    after(() => sub.stopHooks.forEach((hook) => hook()));
+
+    const puzzleId = 'fXchzrh8X9EoSZu6k';
+    const newTagIds = ['NwhNGo64jRs384HwN', '27YauwyRpL6yMsCef'];
+
+    publishJoinedQuery(sub, {
+      model: Puzzles,
+      foreignKeys: [{
+        field: 'tags',
+        join: {
+          model: Tags,
+        },
+      }],
+    }, { _id: puzzleId });
+
+    await Puzzles.updateAsync(puzzleId, { $set: { tags: newTagIds } });
+    // no better way to wait for update to propagate than sleeping
+    await new Promise<void>((r) => { Meteor.setTimeout(r, 50); });
```

Comment:
If this winds up being flaky in CI, a possibly-more-resilient approach might be to poll every 50ms until the condition you wish to observe holds, up to a (much higher than 50ms) timeout.  Happy to cross my fingers for now, but my experience is that losing races tends to be pretty endemic when using shared infrastructure (or in general, where you don't have control over work-scheduling).

Reply:
This prompted me to think about how to word this in a way that doesn't require a timeout, and I realized that since this is an entirely server-side test, I could just set my own `observeChanges` watcher and block until that fires (which indicates that the change has propagated through the oplog)